### PR TITLE
Feature globally germ aware fpr

### PR DESCRIPTION
**Pull request by coreyostrove**
This PR includes a significant number of changes, primarily with regards to the experiment design code, as suggested by the name of the branch, but also with regards to a number of miscellaneous changes that were made in other parts of the codebase to serve projects that had been worked on while using this branch. Here is my best attempt at a summary, broken into categories:

- Per-Germ Global Fiducial pair reduction: implementation of a new version of fiducial pair reduction outlined in [this paper](https://arxiv.org/abs/2308.08781).
- Fiducial selection updates
    - Performance improvements for deduping candidate fiducials for non-clifford gate sets
    - Refactors for candidate fiducial set generation
    - Clean-up line-label handling on empty fiducial
    - Various bugfixes
- Germ selection bugfixes and minor updates:
    - Gate sets with statically parameterized gates
    - Memory efficient implementation of germ list completeness check
    - Better implicit model support/bugfixes
- Fisher information
    - MPI-support
    - Addition of approximate fisher information
    - Option for more memory efficient implementation
    - Additional options for pre-computation/caching
- Miscellaneous
    - Better reporting support for ModelTest protocols w/o gauge-optimization
    - Bugfix in distributed memory implementation of customsolve
    - Error generator plot bugfix when number of qubits>2
    - Add product method to MapForwardSimulator
    - Front load COPA layout construction and associated memory checks
    - Add a new method to StateSpace base class for retreiving a list of state space labels (implemented to address a bug that was in one of the fiducial selection routines, but should be useful more generally).

Unit tests have been added for new experiment design code and some existing coverage gaps filled. Updated Fidsel+Germsel and FPR tutorials to reflect these changes and highlight new capabilities.

### CODE REVIEW

**Comment by coreyostrove** [comment]
Regarding the failing unit tests in the latest runs, looking at the errors I thought to myself 'how much do I want to bet they just released numpy 1.26?'. And lo and behold that is exactly what happened. It dropped on 9/16 and it looks like a few bits of previously deprecated code hit their expiry date.

**Comment by sserita** [comment]
Well gotta love that. How bad are the deprecations? Should we pin 1.25 for this PR, or just bite the bullet and update?

**Comment by sserita** [comment]
Never mind, found the few updates we should need for 1.26. I'll merge that into here once it passes on develop, and then we should be ready for review as normal.

**Comment by coreyostrove** [comment]
You're the best, @sserita! Thanks for taking care of that!

**Comment by sserita** [comment]
Np, thanks for all your hard work on the branch so far. :) I'll take a look before our next dev meeting, and I think Erik had a few questions about what's in here too. But hopefully we can pull this in next week and you'll be off the hook for anything else for the next release (except maybe a review on my MCFE stuff). :P

File: pygsti_new_fpr_requirements_04_24_2023.txt
Comment (sserita) [inline pygsti_new_fpr_requirements_04_24_2023.txt:1]: I don't think we need this file right? Or maybe it should replace the current requirements.txt? The requirements.txt files seem to be used only in the GitHub workflows, although it's not clear to me exactly how necessary they are... Either way, this feels like it should either be removed or merged in somehow.

File: pygsti/algorithms/fiducialpairreduction.py [hunk 1]
```diff
@@ -1413,7 +1417,7 @@ def _get_per_germ_power_fidpairs_greedy(prep_fiducials, meas_fiducials, pre_povm
     #search space.
     printer.log('Removing useless fidcuial pairs with trivial sensitivity to the germ kite parameters. Sensitivity threshold is %.3f'%(sensitivity_threshold), 3)
```
Comment (sserita) [inline pygsti/algorithms/fiducialpairreduction.py:1418]: typo: fiducial

File: pygsti/algorithms/fiducialpairreduction.py [hunk 2]
```diff
@@ -1573,4 +1577,614 @@ def filter_useless_fid_pairs(fiducial_indices, element_map, complete_jacobian, s
...
+                             comm=None, evd_tol=evd_tol,  float_type=float_type,
+                             printer=printer, return_eigs=False, dprobs_dict = dprobs_dict)
+    
+    #It can happen that there are fiducial pairs which are entirely insensitive to any
+    #of the selected amplified parameters, which show up as empty matrices in the list of directional derivatives.
+    #We should identify which fiducial pairs those are and remove them from the 
+    #search space.
+    
+    #TODO: Add back this functionality for the parameter vector based FPR.
+    printer.log('Removing useless fidcuial pairs with trivial sensitivity to the vector of parameters.', 3)
```
Comment (sserita) [inline pygsti/algorithms/fiducialpairreduction.py:1820]: typo: fiducial

File: pygsti/algorithms/fiducialpairreduction.py [hunk 3]
```diff
@@ -1573,4 +1577,614 @@ def filter_useless_fid_pairs(fiducial_indices, element_map, complete_jacobian, s
...
+
+    germ_set_spanning_kwargs : dict, optional (default None)
+        A dictionary of optional kwargs for the function
+        `pygsti.algorithms.germselection.germ_set_spanning_vectors`.
+        See doctring for that function for more details. Only utilized
+        if the germ_vector_spanning_set argument of this function is None.
+    
+    precomputed_jacobians : dict, optional (default None)
+        An optional dictionary of precomputed jacobian dictionaries for the
+        germ fidcuials pair set. The keys are germs and the values are the dictionaries
```
Comment (sserita) [inline pygsti/algorithms/fiducialpairreduction.py:1680]: typo: fiducial

File: pygsti/algorithms/germselection.py [hunk 4]
```diff
@@ -4285,3 +4369,597 @@ def stable_pinv(mat):
...
+#---------- Minimal Germ Spanning Vectors------------#
+#Not a great name for this section of code. Idea here is to
+#take a user-inputed AC germ-set (presumably from a prior germ
+#selection run and then identify a minimal set of amplified directions
+#that spans model space collectively accross all germs.
+#this can then be used as input to a version of FPR that is
+#globally aware of the overlap in amplified directions
+#of parameter space.
+
+def germ_set_spanning_vectors(target_model, germ_list, assume_real=False, float_type=_np.cdouble,
```
Comment (sserita) [inline pygsti/algorithms/germselection.py:4383]: This appears to be defined here but only used in the FPR code. Thoughts on moving it into FPR?

**Review by sserita — COMMENTED** [review]
Thanks for the great work as usual @coreyostrove. I have two main comments for you to address about the requirements.txt and the location of germ_set_spanning_vectors - the answer to both could be "no its fine, plz just approve". But I've also got a few minor typos to fix while it's bounced back to you... sorry for being annoying about those. :P

File: pygsti_new_fpr_requirements_04_24_2023.txt
Comment (coreyostrove) [reply pygsti_new_fpr_requirements_04_24_2023.txt:1]: This was committed by accident (I was transferring a local environment for another project tied to this branch and forgot I had this in the folder). Thanks for catching that!

File: pygsti/algorithms/germselection.py (hunk 4 above)
Comment (coreyostrove) [reply pygsti/algorithms/germselection.py:4383]: Good question. I bounced back and forth between putting this in the germ selection code or the FPR code before settling on germ selection for two main reasons. The first is that algorithmically this function shares much more in common with the algorithms used in germ selection, and uses a lot of the same helper functions which are defined in that file and are conversely not used in the fiducial pair reduction algorithms (so there'd be a lot of imports in the FPR code for functions only used in this context). The other was that in the long term I have applications in mind for this functionality that aren't directly related to fiducial pair reduction, so it was probably always going to have an imperfect home either way. I'm not deeply wedded to this though.

Comment (sserita) [reply pygsti/algorithms/germselection.py:4383]: Ok, that's good enough for me. Just wanted to make sure this was an intentional decision :)

**Pull request merged**